### ISSUE TITLE
Infrastructure todos

### ISSUE THREAD

**Issue by simoncozens** [fonttools/fontspector#4]
The spike has been fun but it's time to get serious. A few things need to be nailed down *now* or else people will start porting loads of checks and it will be a nightmare to change things around later:

- [ ]  *Source fixing*: associate testables with their sources and fix a few things at source. Since fonts can have different source formats, this is going to require some infrastructure to handle cleanly.
- [x] *Robust error handling*: Nothing should `unwrap`, ever. Make sure that *anything* that can fail returns a `Result` with a sensible error type (no more `Result<(),String>`), and hand back `Result`s at reasonable boundaries.
- [x] *Python bridge plugin*: We don't need it ourselves, but do it anyway to make sure it can be done.
- [x] *Reporter backends*: Templatable backends sharing code between terminal/MD/HTML.
- [x] *Profile/configuration*: Make sure that checks can get access to any contextual information they need from the user or profile.
- [x] *Caching*: Currently we're firing up a new `FontRef` every time we want something from the font - is there a better way? (Note that we're also firing up a new `TestFont` for each testable every time we do a check, so optimising `TestFont` won't help you.)
- [x] *Naming*: "Fontspector" is ugly, will probably steal the [fontrefinery](https://github.com/simoncozens/fontrefinery) name for this.

**Comment by davelab6** [comment]
> steal the [fontrefinery](https://github.com/simoncozens/fontrefinery) name for this.

That is a _refined_ name indeed, suitable for the _industrial scale_ it is put to use on :D

Please take a moment to fix up the project to follow the Google Open Source Project Office boilerplate stuff, AUTHORS.txt, all that jazz

**Comment by felipesanches** [comment]
> > steal the [fontrefinery](https://github.com/simoncozens/fontrefinery) name for this.
> 
> That is a _refined_ name indeed, suitable for the _industrial scale_ it is put to use on :D

I'm hoping this project will become FontBakery itself in the future (long-term), actually replacing the code at the https://github.com/fonttools/fontbakery repo (and not having python-based releases anymore). In the mean-time, I think it is fine to keep "Fontspector" as a code-name during its prototyping / experimental stage.

**Comment by simoncozens** [comment]
> Currently we're firing up a new FontRef every time we want something from the font - is there a better way?

For future readers: It *is* possible to cache parsed and owned font information (I made a branch called `experimental-cache` which does this), but this doesn't seem to give any performance improvement - presumably because of cache overhead and lock contention. Parsing the whole font again seems to be quick enough, provided you avoid e.g. referring to `font().charmap()` inside a loop.

**Comment by simoncozens** [comment]
> I'm hoping this project will become FontBakery itself in the future (long-term)

I agree with this; at the same time, I think it's probable we will want to have both projects in production in different places.
